src/universal/helpers/bag.ts

```diff
@@ -2,68 +2,84 @@ import { LatLngLiteral, LatLngTuple } from 'leaflet';
 
 import { BAGQueryParams, BAGAdreseerbaarObject } from '../types/bag';
 
-export function extractAddress(rawAddress: string): BAGQueryParams {
-  // Remove everything but alphanumeric, dash, dot, apostrophe and space.
-  const address = rawAddress.replace(/[^/'0-9-.\s\p{Script=Latin}+]/giu, '');
+type ExtractUtils = {
+  pattern: RegExp;
+  formatter?: (s: string) => unknown;
+};
+
+// The order here is important since we delete our found pattern to make it easier for the next.
+const patterns: Partial<Record<keyof BAGQueryParams, ExtractUtils>> = {
+  postcode: {
+    pattern: /[1-9][0-9]{3} ?(?!sa|sd|ss)[a-z]{2}/i,
+    // Postcode can have a space between numbers and letters.
+    formatter: (postcode: string) => postcode.replace(' ', ''),
+  },
+  openbareruimteNaam: {
+    // @ts-ignore unicode does works
+    pattern: /(?<![0-9a-z])([0-9]e)?[a-z_.'/\-\p{L} ]+(?![0-9a-z])/iu,
+  },
+  huisnummer: {
+    pattern: /\d+/i,
+    formatter: (huisnummer) => parseInt(huisnummer),
+  },
+  huisnummertoevoeging: { pattern: /\b\w{1,5}\b/i },
+};
+
+/** Extract an address from free form input.
+
+ * @throws {Error}
+ */
+export function extractAddress(rawText: string): BAGQueryParams {
+  if (!rawText) {
+    throw `Cannot extract address out of bad input: '${rawText}'`;
+  }
 
-  const words = [];
-  const s = address.split(' ');
+  const cleanText = rawText
+    .trim()
+    // Remove everything but alphanumeric, dash, dot, apostrophe and space.
+    // @ts-ignore unicode does work
+    .replace(/[^/'0-9-.\s\p{Script=Latin}+]/giu, '')
+    // Spaces to single space.
+    .replace(/\s{2,}/g, ' ');
 
-  if (s.length < 2) {
-    throw Error(
-      `Address should consist of minimally two parts. A streetname and a housenumber.
-      Address: '${address}'`
-    );
+  const [, result] = Object.entries(patterns).reduce(extract, [cleanText, {}]);
+
+  if (!(result.openbareruimteNaam || result.postcode)) {
+    throw `Cannot execute a BAG query without a openbareruimtenaam (streetname) or postcode.
+    Error resulted from input: '${rawText}' and got parsed into:
+    ${JSON.stringify(result, null, 2)}`;
   }
 
-  let i = 0;
-  const onDigit = new RegExp(/\d/);
+  return result;
+}
 
-  for (; i < s.length; i++) {
-    const word = s[i];
-    if (word[0].match(onDigit)) {
-      // The first housenumber found, so there are no more streetname words left.
-      break;
-    }
-    words.push(word);
+type Text = string;
+
+function extract(
+  acc: [Text, BAGQueryParams],
+  namedPattern: [string, ExtractUtils]
+): [Text, BAGQueryParams] {
+  const [name, utils] = namedPattern;
+  const { pattern, formatter } = utils;
```

Comment:
Deze kan je combineren toch?
```suggestion
  const [name, { pattern, formatter }] = namedPattern;
```

Reply:
Goede, dat is wel iets cleaner